core/filelocation.cpp

```diff
@@ -0,0 +1,235 @@
+{
+	switch (type) {
+	default:
+	case NONE:
+		return;
+	case LOCAL_FILE:
+		name = QFile::encodeName(QDir::toNativeSeparators(name));
+		return;
+	case CLOUD_GIT:
+	case CLOUD_GIT_OFFLINE:
+	case GIT:
+		if (name.startsWith("file://", Qt::CaseInsensitive)) {
+			// Treat "file://..." URLs as local repositories.
+			name = name.mid(7);
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(!isRemoteRepository(name)) {
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(name.startsWith("https://", Qt::CaseInsensitive)) {
+			int at = name.indexOf('@');
+			int slash = name.indexOf('/', 8);
+			if (at >= 0 && slash >= 0 && slash > at) {
+				user = name.mid(8, at - 8);
+				name.remove(8, at - 7);
+			}
+		}
+	}
+}
+
+// Check if filename is of the form "repository[branch]" or "repository[branch].ssrf".
+static bool isGitFilename(const QString &name)
+{
+	QString trimmed = name.trimmed();
+	if (trimmed.endsWith(".ssrf", Qt::CaseInsensitive))
+		trimmed.truncate(trimmed.count() - 5);
+	return trimmed.right(1) == "]" &&
+	       trimmed.lastIndexOf('[') > 0;
+}
+
+// Returns (repository, branch) pair.
+static std::pair<QString, QString> parseGitFilename(const QString &fn)
+{
+	if (!isGitFilename(fn))							// Oops. Filename doesn't follow the rules. Leave branch empty.
+		return std::make_pair(fn, QString());
+
+	int branch_from = fn.lastIndexOf('[');
+	int branch_to = fn.lastIndexOf(']');
+	QString repository = fn.left(branch_from);				// Guaranteed OK owing to isGitFilename() == true
+	QString branch = fn.mid(branch_from + 1, branch_to - branch_from - 1);	// ditto
+	return std::make_pair(repository, branch);
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_, const QString &branch_) : type(type_), name(name_), branch(branch_)
+{
+	if (type == NONE)
+		return;
+	fixName();
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_) : type(type_)
+{
+	if (type == NONE)
+		return;
+	if (type == LOCAL_FILE) {
+		name = name_;
+	} else {
+		// Is a git repository
+		std::pair<QString, QString> repo_branch = parseGitFilename(name_);
+		name = repo_branch.first;
+		branch = repo_branch.second;
+	}
+	fixName();
+}
+
+FileLocation::FileLocation(const QString &type_, const QString &name_, const QString &branch_) : name(name_), branch(branch_)
+{
+	if (type_ == "none")
+		type = NONE;
+	else if (type_ == "local")
+		type = LOCAL_FILE;
+	else if (type_ == "cloud")
+		type = CLOUD_GIT;
+	else if (type_ == "cloud_offline")
+		type = CLOUD_GIT_OFFLINE;
+	else if (type_ == "git")
+		type = GIT;
+	else
+		*this = guessFromFileName(name_); // Unknown or empty type -> try to guess from filename
+	fixName();
+}
+
+FileLocation FileLocation::guessFromFileName(const QString &name)
+{
+	if (name.simplified().isEmpty()) {
+		return FileLocation();
+	} else if (isGitFilename(name)) {
+		std::pair<QString, QString> repo_branch = parseGitFilename(name);
+		// TODO: Check for git repository if local!!
```

Comment:
remember, ALL git repositories are at least in part local - even cloud

Reply:
Yes, I realize that.

The function you're looking at is `guessFromFileName`. Once there is an "load/save git repository" menu entry, this function will only be used for filenames provided on the command line. So the idea of this TODO is to treat things like "xxx[yyy]" as files if this is not branch "yyy" in the local repository "xxx".